File: vane-core/src/main/java/org/oddlama/vane/core/config/ConfigMaterialSetField.java
package org.oddlama.vane.core.config;

import static org.oddlama.vane.util.MaterialUtil.materialFrom;
import static org.oddlama.vane.util.StorageUtil.namespacedKey;

import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Function;
import org.bstats.bukkit.Metrics;
import org.bstats.charts.AdvancedPie;
import org.bukkit.Material;
import org.bukkit.configuration.file.YamlConfiguration;
import org.oddlama.vane.annotation.config.ConfigMaterialSet;
import org.oddlama.vane.core.YamlLoadException;

public class ConfigMaterialSetField extends ConfigField<Set<Material>> {

    public ConfigMaterialSet annotation;

    public ConfigMaterialSetField(
        Object owner,
        Field field,
        Function<String, String> mapName,
        ConfigMaterialSet annotation
    ) {
        super(owner, field, mapName, "set of materials", annotation.desc());
        this.annotation = annotation;
    }

    private void appendMaterialSetDefinition(
        StringBuilder builder,
        String indent,
        String prefix,
        Set<Material> def
    ) {
        appendListDefinition(builder, indent, prefix, def, (b, m) -> {
            b.append("\"");
            b.append(escapeYaml(m.getKey().getNamespace()));
            b.append(":");
            b.append(escapeYaml(m.getKey().getKey()));
            b.append("\"");
        });
    }

    @Override
    public Set<Material> def() {
        final var override = overriddenDef();
        if (override != null) {
            return override;
        } else {
            return new HashSet<>(Arrays.asList(annotation.def()));
        }
    }

    @Override
    public boolean metrics() {
        final var override = overriddenMetrics();
        if (override != null) {
            return override;
        } else {
            return annotation.metrics();
        }
    }

    @Override
    public void registerMetrics(Metrics metrics) {
        if (!this.metrics()) return;
        metrics.addCustomChart(
            new AdvancedPie(yamlPath(), () -> {
                final var values = new HashMap<String, Integer>();
                for (final var v : get()) {
                    values.put(v.getKey().toString(), 1);
                }
                return values;
            })
        );
    }

    @Override
    public void generateYaml(StringBuilder builder, String indent, YamlConfiguration existingCompatibleConfig) {
        appendDescription(builder, indent);

        // Default
        builder.append(indent);
        builder.append("# Default:\n");
        appendMaterialSetDefinition(builder, indent, "# ", def());

        // Definition
        builder.append(indent);
        builder.append(basename());
        builder.append(":\n");
        final var def = existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath())
            ? loadFromYaml(existingCompatibleConfig)
            : def();
        appendMaterialSetDefinition(builder, indent, "", def);
    }

    @Override
    public void checkLoadable(YamlConfiguration yaml) throws YamlLoadException {
        checkYamlPath(yaml);

        if (!yaml.isList(yamlPath())) {
            throw new YamlLoadException("Invalid type for yaml path '" + yamlPath() + "', expected list");
        }

        for (var obj : yaml.getList(yamlPath())) {
            if (!(obj instanceof String)) {
                throw new YamlLoadException("Invalid type for yaml path '" + yamlPath() + "', expected string");
            }

            final var str = (String) obj;
            final var split = str.split(":");
            if (split.length != 2) {
                throw new YamlLoadException(
                    "Invalid material entry in list '" + yamlPath() + "': '" + str + "' is not a valid namespaced key"
                );
            }

            final var mat = materialFrom(namespacedKey(split[0], split[1]));
            if (mat == null) {
                throw new YamlLoadException(
                    "Invalid material entry in list '" + yamlPath() + "': '" + str + "' does not exist"
                );
            }
        }
    }

    public Set<Material> loadFromYaml(YamlConfiguration yaml) {
        final var set = new HashSet<Material>();
        for (var obj : yaml.getList(yamlPath())) {
            final var split = ((String) obj).split(":");
            set.add(materialFrom(namespacedKey(split[0], split[1])));
        }
        return set;
    }

    public void load(YamlConfiguration yaml) {
        try {
            field.set(owner, loadFromYaml(yaml));
        } catch (IllegalAccessException e) {
            throw new RuntimeException("Invalid field access on '" + field.getName() + "'. This is a bug.");
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/config/ConfigMaterialSetField.kt
package org.oddlama.vane.core.config

import org.bstats.bukkit.Metrics
import org.bstats.charts.AdvancedPie
import org.bukkit.Material
import org.bukkit.configuration.file.YamlConfiguration
import org.oddlama.vane.annotation.config.ConfigMaterialSet
import org.oddlama.vane.core.YamlLoadException
import org.oddlama.vane.util.MaterialUtil.materialFrom
import org.oddlama.vane.util.StorageUtil.namespacedKey
import java.lang.reflect.Field

/**
 * Config field handler for material set values.
 *
 * @param owner object containing the reflected config field.
 * @param field reflected config field.
 * @param mapName maps Java field names to YAML paths.
 * @param annotation source annotation metadata.
 */
class ConfigMaterialSetField(
    owner: Any?,
    field: Field,
    mapName: (String?) -> String?,
    /** Annotation metadata for this field. */
    var annotation: ConfigMaterialSet
) : ConfigField<MutableSet<Material>?>(owner, field, mapName, "set of materials", annotation.desc) {

    /** Formats a material as a quoted namespaced key. */
    private fun Material.keyString() = "\"${escapeYaml(key.namespace)}:${escapeYaml(key.key)}\""

    /** Appends a material set definition block. */
    private fun appendMaterialSetDefinition(builder: StringBuilder?, indent: String?, prefix: String?, def: MutableSet<Material>) {
        appendListDefinition(builder, indent, prefix, def) { b, m -> b!!.append(m!!.keyString()) }
    }

    /** Returns the default value for this config field. */
    override fun def(): MutableSet<Material> = overriddenDef() ?: annotation.def.toMutableSet()

    /** Returns whether metrics collection is enabled for this field. */
    override fun metrics(): Boolean = overriddenMetrics() ?: annotation.metrics

    /** Registers an advanced pie metric containing selected material keys. */
    override fun registerMetrics(metrics: Metrics?) {
        if (metrics == null || !this.metrics()) return
        metrics.addCustomChart(AdvancedPie(yamlPath()) {
            get()!!.associate { it.key.toString() to 1 }
        })
    }

    /** Generates YAML for this field. */
    override fun generateYaml(builder: StringBuilder, indent: String, existingCompatibleConfig: YamlConfiguration?) {
        appendDescription(builder, indent)
        builder.append("$indent# Default:\n")
        appendMaterialSetDefinition(builder, indent, "# ", def())
        builder.append("$indent${basename()}:\n")
        val def: MutableSet<Material> = if (existingCompatibleConfig != null && existingCompatibleConfig.contains(yamlPath()))
            loadFromYaml(existingCompatibleConfig) else def()
        appendMaterialSetDefinition(builder, indent, "", def)
    }

    /** Validates that this field is loadable from YAML. */
    @Throws(YamlLoadException::class)
    override fun checkLoadable(yaml: YamlConfiguration) {
        checkYamlPath(yaml)
        if (!yaml.isList(yamlPath()))
            throw YamlLoadException("Invalid type for yaml path '${yamlPath()}', expected list")
        for (obj in yaml.getList(yamlPath())!!) {
            if (obj !is String)
                throw YamlLoadException("Invalid type for yaml path '${yamlPath()}', expected string")
            val (ns, key) = requireNamespacedKeyParts(yamlPath(), obj, "material")
            materialFrom(namespacedKey(ns, key))
                ?: throw YamlLoadException("Invalid material entry in list '${yamlPath()}': '$obj' does not exist")
        }
    }

    /** Loads this field value from YAML. */
    fun loadFromYaml(yaml: YamlConfiguration): MutableSet<Material> =
        yaml.getList(yamlPath())!!.mapTo(mutableSetOf()) { obj ->
            val (ns, key) = requireNamespacedKeyParts(yamlPath(), obj as String, "material")
            materialFrom(namespacedKey(ns, key))!!
        }

    /** Writes the loaded value into the reflected field. */
    override fun load(yaml: YamlConfiguration) {
        try {
            field.set(owner, loadFromYaml(yaml))
        } catch (_: IllegalAccessException) {
            throw RuntimeException("Invalid field access on '${field.name}'. This is a bug.")
        }
    }
}
